awx/api/views/__init__.py

```diff
@@ -3826,6 +3826,168 @@ def get_queryset(self):
+        (4, 'f4194f14-e406-4124-8519-0fdb08b18f4b', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (5, '39f7ad99-dbf3-41e0-93f8-9999db4004f2', 'aae0d189-e3cb-102a-9f00-00000000000c')
+        (6, 'aae0d189-e3cb-102a-9f00-000000000008', 'aae0d189-e3cb-102a-9f00-000000000006')
+        (7, '39a49992-5ca4-4b6c-b178-e56d0b0333da', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (8, '504f3b28-3ea8-4f6f-bd82-60cf8e807cc0', 'aae0d189-e3cb-102a-9f00-000000000008')
+        (9, 'a242be54-ebe6-4021-afab-f2878bff2e9f', '4598d19e-93b4-4e33-a0ae-b387a7348964')
+
+    output:
+        {
+        "1": {
+            "rowNumber": 0,
+            "numChildren": 8
+        },
+        "2": {
+            "rowNumber": 1,
+            "numChildren": 6
+        },
+        "3": {
+            "rowNumber": 2,
+            "numChildren": 2
+        },
+        "6": {
+            "rowNumber": 5,
+            "numChildren": 2
+        }
+        }
+        "meta_event_nested_parent_uuid": {}
+        }
+    counter 1 is event 0, and has 8 children
+    counter 2 is event 1, and has 6 children
+    etc.
+
+    The UI also needs to be able to collapse over "meta" events -- events that
+    show up due to verbosity or warnings from the system while the play is running.
+    These events have a 0 level event, with no parent uuid.
+
+        playbook_on_start
+        verbose
+          playbook_on_play_start
+            playbook_on_task_start
+              runner_on_start        <- level 3
+        verbose                      <- jump to level 0
+        verbose
+                runner_on_ok         <- jump back to level 3
+              playbook_on_task_start
+                runner_on_start
+                runner_on_ok
+        verbose
+        verbose
+              playbook_on_stats
+
+    These verbose statements that fall in the middle of a series of children events
+    are problematic for the UI.
+    To help, this view will attempt to place the events into the hierarchy, without
+    the event level jumps.
+
+        playbook_on_start
+          verbose
+          playbook_on_play_start
+            playbook_on_task_start
+              runner_on_start        <- A
+                verbose              <- this maps to the uuid of A
+                verbose
+                runner_on_ok
+              playbook_on_task_start <- B
+                runner_on_start
+                runner_on_ok
+                verbose              <- this maps to the uuid of B
+                verbose
+              playbook_on_stats
+
+    The output will include a JSON object where the key is the event counter,
+    and the value is the assigned nested uuid.
+    '''
+
+    renderer_classes = [JSONRenderer]
+    meta_events = ('debug', 'verbose', 'warning', 'error', 'system_warning', 'deprecated')
+
+    def get(self, request, **kwargs):
+        job = get_object_or_404(models.Job, pk=kwargs['pk'])
+        if not job.event_processing_finished:
+            return Response({})
```

Comment:
If this won't run when event processing isn't finished, then why do we need to bother with the rowNumber? It should be the same as counter minus 1.

Reply:
There can't reasonably be a gap when `job.event_processing_finished` is True. Consider the logic from that.

https://github.com/ansible/awx/blob/6c1adade250020fcf900ecfc6d33e6f8274c8e0a/awx/main/models/unified_jobs.py#L1022

The way this works:
 - Control process receives output from receptor, and it does minimal processing of this and submits the data to receptor
   - This process counts the number of events that it sends, and saves that to `emitted_events`
 - Callback receiver pulls the data off the redis queue and then saves it to database

If `event_processing_finished` never transitions to `True`, then that's an API bug. Doesn't matter what type of job or what it does. This is accounting - the callback receiver should always save the same number of events that the control process submitted.

If the job is actually _running_ then this endpoint returns `{}` anyway